File: 2065f_sas.cpp
/**
 * @file 2065f_sas.cpp
 * @author Ukeme Edet (ukemeedet2207@gmail.com)
 * @brief The solution to the problem Skibidus and
 * Slay(https://codeforces.com/contest/2065/problem/F)
 * @version 0.1
 * @date 2025-03-19
 *
 * @copyright Copyright (c) 2025
 *
 */

#include <bits/stdc++.h>

#define ll long long

using namespace std;

/**
 * @brief The main function
 *
 * @return int
 */
int main(void) {
    ios::sync_with_stdio(0);
    cin.tie(0);
    ll t, n, u, v, i, j;
    vector<ll> a;
    vector<char> res;
    vector<vector<ll>> adj;
    map<ll, ll> fm;

    cin >> t;
    while (t--) {
        cin >> n;
        adj = {}, res = {};
        adj.resize(n + 1), res.resize(n, '0'), a.resize(n);
        for (i = 0; i < n; i++)
            cin >> a[i];
        for (i = 0; i < n - 1; i++) {
            cin >> u >> v;
            adj[u].push_back(v);
            adj[v].push_back(u);
        }
        for (i = 1; i <= n; i++) {
            fm = {};
            for (j = 0; j < adj[i].size(); j++)
                fm[a[adj[i][j] - 1]]++;
            fm[a[i - 1]]++;
            for (const auto &[k, v] : fm)
                if (v > 1)
                    res[k - 1] = '1';
        }
        for (i = 0; i < res.size(); i++)
            cout << res[i];
        cout << '\n';
    }
    return (EXIT_SUCCESS);
}

File: 2065f_sas.py
"""
@file: 2065f_sas.py
@author: Ukeme Edet (ukemeedet2207@gmail.com)
@brief: The solution to the problem Skibidus and Slay(https://codeforces.com/contest/2065/problem/F)
@version: 1.0
@date: 2025-03-19

@copyright: Copyright (c) 2025
"""

from collections import defaultdict
from sys import stdin, stdout
from typing import Counter


def input():
    return stdin.readline().strip()


def print(*args, sep=" ", end="\n"):
    return stdout.write(sep.join(str(x) for x in args) + end)


def main():
    for _ in range(int(input())):
        n = int(input())
        a = [int(x) for x in input().split()]
        ch = [[] for _ in range(n + 1)]
        for i in range(n - 1):
            u, v = [int(x) for x in input().split()]
            ch[u].append(v)
            ch[v].append(u)
        res = ["0"] * n
        for i in range(1, n + 1):
            fm = defaultdict(int)
            for j in ch[i]:
                fm[str(a[j - 1])] += 1
            fm[str(a[i - 1])] += 1
            for k in fm.keys():
                if fm[k] > 1:
                    res[int(k) - 1] = "1"
        print("".join(res))


if __name__ == "__main__":
    main()
